modules/images/webp-uploads/load.php

```diff
@@ -683,3 +683,190 @@ function webp_uploads_update_rest_attachment( WP_REST_Response $response, WP_Pos
+ * right after an edit has taken place and the current sources are no longer accurate.
+ *
+ * @since n.e.x.t
+ *
+ * @param int   $attachment_id The ID representing the attachment.
+ * @param array $data The current metadata of the attachment.
+ * @return array The updated metadata for the attachment.
+ */
+function webp_uploads_backup_sources( $attachment_id, $data ) {
+	if ( ! isset( $data['sources'] ) ) {
+		return $data;
+	}
+
+	$data['_sources'] = $data['sources'];
+	// Remove the current sources as at this point the current values are no longer accurate.
+	unset( $data['sources'] );
+
+	return $data;
+}
+
+/**
+ * Restore an image from the backup sizes, the current hook moves the `sources` from the `full-orig` key into
+ * the top level `sources` into the metadata, in order to ensure the restore process has a reference to the right
+ * images. When `IMAGE_EDIT_OVERWRITE` is defined and is truthy the function would loop into the sources array
+ * and remove any sources that was created as an edited version, the filenames for those cases are constructed out
+ * of `e-{digits}` where `{digits}` is a number of length 13.
+ *
+ * @param int   $attachment_id The ID of the attachment.
+ * @param array $data The current metadata to be stored in the attachment.
+ * @return array The updated metadata of the attachment.
+ */
+function webp_uploads_restore_image( $attachment_id, $data ) {
+	$backup_sizes = get_post_meta( $attachment_id, '_wp_attachment_backup_sizes', true );
+
+	if ( ! is_array( $backup_sizes ) ) {
+		return $data;
+	}
+
+	// If `IMAGE_EDIT_OVERWRITE` is defined and is truthy remove any edited images if present before replacing the metadata.
+	if ( defined( 'IMAGE_EDIT_OVERWRITE' ) && IMAGE_EDIT_OVERWRITE ) {
+		$file     = get_attached_file( $attachment_id );
+		$dirname  = pathinfo( $file, PATHINFO_DIRNAME );
+		$metadata = wp_get_attachment_metadata( $attachment_id );
+
+		$sources = isset( $metadata['sources'] ) && is_array( $metadata['sources'] ) ? $metadata['sources'] : array();
+
+		foreach ( $sources as $mime => $properties ) {
+			if ( empty( $properties['file'] ) ) {
+				continue;
+			}
+
+			// Delete only if it's an edited image.
+			if ( preg_match( '/-e\d{13}/', $properties['file'] ) ) {
+				$delete_file = path_join( $dirname, $properties['file'] );
+				wp_delete_file( $delete_file );
+			}
+		}
+
+		foreach ( $metadata['sizes'] as $size_name => $properties ) {
+			if ( ! isset( $properties['sources'] ) || ! is_array( $properties['sources'] ) ) {
+				continue;
+			}
+
+			foreach ( $properties['sources'] as $mime => $source_properties ) {
+				if ( empty( $source_properties['file'] ) ) {
+					continue;
+				}
+
+				// Delete only if it's an edited image.
+				if ( preg_match( '/-e\d{13}/', $source_properties['file'] ) ) {
+					$delete_file = path_join( $dirname, $source_properties['file'] );
+					wp_delete_file( $delete_file );
+				}
+			}
+		}
```

Comment:
I actually think this entire block of code can be removed. We don't support generating additional MIME types for image edits yet, so none of the edited images (neither the original nor any `sizes`) will have any `sources` that we'd need to delete.

This will all be part of #158.

Reply:
> I actually think this entire block of code can be removed. We don't support generating additional MIME types for image edits yet, so none of the edited images (neither the original nor any sizes) will have any sources that we'd need to delete.

That's correct we currently don't have support for that but the existing issue #158 only handles the creation of those edited version images, this logic already covers the fact that you edited an image and then you restored, I agree that this logic right now won't be executed since we need #158 completed, however, this already works we can either move this code into a separate PR that's going to be related with #228 and #158 or we can create one to keep track of that, let me know what do you prefer but to me it seems reasonable to include it here already as it's part of the restore logic from #228.

Here's a test case covering that scenario: 

- https://github.com/WordPress/performance/blob/4820501b979d4e710baed19bd75d4597b628562b/tests/modules/images/webp-uploads/webp-uploads-test.php#L988

Reply:
I think it would be better to move this out of here, to keep it simple. It makes more sense to have another PR against #158 for that IMO, since it is only needed if we also implement the other bit of #158. Including it here is a bit preliminary.

Reply:
Sounds good just removed this section of code and added a `TODO` so we can address it once #158 is completed.